chain/chunks/src/adapter.rs

```diff
@@ -0,0 +1,202 @@
+        self.do_send(ShardsManagerRequest::ProcessPartialEncodedChunk(partial_encoded_chunk));
+    }
+    fn process_partial_encoded_chunk_forward(
+        &self,
+        partial_encoded_chunk_forward: PartialEncodedChunkForwardMsg,
+    ) {
+        self.do_send(ShardsManagerRequest::ProcessPartialEncodedChunkForward(
+            partial_encoded_chunk_forward,
+        ));
+    }
+    fn process_partial_encoded_chunk_response(
+        &self,
+        partial_encoded_chunk_response: PartialEncodedChunkResponseMsg,
+        received_time: Instant,
+    ) {
+        self.do_send(ShardsManagerRequest::ProcessPartialEncodedChunkResponse {
+            partial_encoded_chunk_response,
+            received_time,
+        });
+    }
+    fn process_partial_encoded_chunk_request(
+        &self,
+        partial_encoded_chunk_request: PartialEncodedChunkRequestMsg,
+        route_back: CryptoHash,
+    ) {
+        self.do_send(ShardsManagerRequest::ProcessPartialEncodedChunkRequest {
+            partial_encoded_chunk_request,
+            route_back,
+        });
+    }
+    fn process_chunk_header_from_block(&self, chunk_header: &ShardChunkHeader) {
+        self.do_send(ShardsManagerRequest::ProcessChunkHeaderFromBlock(chunk_header.clone()));
+    }
+    fn update_chain_heads(&self, head: Tip, header_head: Tip) {
+        self.do_send(ShardsManagerRequest::UpdateChainHeads { head, header_head });
+    }
+    fn distribute_encoded_chunk(
+        &self,
+        partial_chunk: PartialEncodedChunk,
+        encoded_chunk: EncodedShardChunk,
+        merkle_paths: Vec<MerklePath>,
+        outgoing_receipts: Vec<Receipt>,
+    ) {
+        self.do_send(ShardsManagerRequest::DistributeEncodedChunk {
+            partial_chunk,
+            encoded_chunk,
+            merkle_paths,
+            outgoing_receipts,
+        });
+    }
+    fn request_chunks(&self, chunks_to_request: Vec<ShardChunkHeader>, prev_hash: CryptoHash) {
+        self.do_send(ShardsManagerRequest::RequestChunks { chunks_to_request, prev_hash });
+    }
+    fn request_chunks_for_orphan(
+        &self,
+        chunks_to_request: Vec<ShardChunkHeader>,
+        epoch_id: EpochId,
+        ancestor_hash: CryptoHash,
+    ) {
+        self.do_send(ShardsManagerRequest::RequestChunksForOrphan {
+            chunks_to_request,
+            epoch_id,
+            ancestor_hash,
+        });
+    }
+    fn check_incomplete_chunks(&self, prev_block_hash: CryptoHash) {
+        self.do_send(ShardsManagerRequest::CheckIncompleteChunks(prev_block_hash));
+    }
+}
+
+pub struct ShardsManagerAdapterAsAdapterForNetwork {
+    pub adapter: Arc<dyn ShardsManagerAdapter>,
+}
+
+impl ShardsManagerAdapterAsAdapterForNetwork {
+    pub fn new(adapter: Arc<dyn ShardsManagerAdapter>) -> Self {
+        Self { adapter }
+    }
+}
+
+impl ShardsManagerAdapterForNetwork for ShardsManagerAdapterAsAdapterForNetwork {
```

Comment:
Does ShardsManagerAdapterForNetwork need to be a trait? Can you just implement ShardsManagerForNetwork as a struct and avoid having both ShardsManagerAdapterForNetwork and ShardsManagerAdapterAsAdapterForNetwork? I think having too many adapters makes the code convoluted unnecessarily.

Reply:
I would advocate for just making network use ShardsManagerAdapter directly instead of create a new adapter for it. The advantage of creating a different interface for network is to restrict the proper usage of network code of ShardsManager. The disadvantage is the code gets too complicated with some many adapters. Since the APIs and how it can be used at call sites are very simple, I don't think there is much risk of network code misusing ShardsManager. I  would prefer to have simpler code here.

Reply:
Yeaaaaa... the proliferation of adapters is seriously confusing. Well, the immediate problem with having network use ShardsManagerAdapter is that near-chunks depends on near-network, so that would create a circular dependency. It can be avoided by moving ShardsManagerAdapter (or perhaps PeerManagerAdapter) into a lower crate, but I'm not sure where that should be - doesn't feel like it should be in primitives - maybe we don't have such a crate today.

Reply:
Ah I see. That is indeed annoying. What if we create two adapters, ShardsManagerAdapterForNetwork and ShardsManagerAdapterForClient and implement 
```
impl<A: MsgRecipient<ShardsManagerRequest>> ShardsManageForNetwork for A
```
and 
```
impl<A: MsgRecipient<ShardsManagerRequest>> ShardsManagerForClient for A
````

Reply:
So... as we discussed offline, I really like this suggestion, and I think this is the direction we want to go. But I ran into some rather subtle roadblocks when trying to make this work.

The most problematic issue is that for testing (at least the current framework), it would be convenient to have a single interface that supports both of these adapter interfaces. For that I need a third trait ShardsManagerAdapterForTest that has both of these traits as bounds, and have the A:MsgRecipient implement ShardsManagerAdapterForTest as well - all of that is fine until I need to take a Arc<dyn ShardsManagerAdapterForTest> and convert it to an Arc<dyn ShardsManagerAdapterForClient> - I was surprised that this is actually an experimental feature in Rust.

Hmm, I don't think this is going to be easy to solve within this PR, so let me try a couple more things, but in the meantime let's just defer this.

Reply:
Actually, I've pushed a revision where I found a workaround to this. Let me know what you think. It's basically what you suggested, plus the following:
 - ShardsManagerAdapterForTest that I mentioned above, plus for_client and for_network to help with trait upcasting
 - I also split the Request enum. This is due to a subtle technicality too, because there's no way to do `impl<A: MsgRecipient<ShardsManagerRequest>> ShardsManagerAdapterForNetwork for A` in the near-chunks crate, because none of these traits belong to near-chunks. So, I have to do that impl in the near-network crate, but then I need the ShardsManagerRequest to also be referenceable from that crate, and therefore I moved that part of the request enum to that crate. This doesn't have much of an effect - what this means is basically that ShardsManager now has handle_client_request and handle_network_request. Nothing fancy.